src/components/Toggle/Toggle.tsx

```diff
@@ -0,0 +1,116 @@
+import { Switch } from '@headlessui/react';
+import clsx from 'clsx';
+import React from 'react';
+import styles from './Toggle.module.css';
+
+type ToggleLabelProps = {
+  /**
+   * Visible text label for the toggle.
+   */
+  children: React.ReactNode;
+  /**
+   * Additional classnames passed in for styling the toggle label.
+   */
+  className?: string;
+};
+
+type ToggleButtonProps = {
+  /**
+   * Additional classnames passed in for styling the toggle button.
+   */
+  className?: string;
+  /**
+   * Whether toggle is checked or unchecked.
+   */
+  checked: boolean;
+  /**
+   * Callback called when toggle state is changed.
+   */
+  onChange: (checked: boolean) => void;
+  /**
+   * Whether toggle is disabled.
+   */
+  disabled?: boolean;
+};
+
+type ToggleProps = ToggleButtonProps & {
+  /**
+   * When possible, use a visible label through the `label` prop instead.
+   * In rare cases where there's no visible label, you must provide an
+   * `aria-label` for screen readers.
+   */
+  'aria-label'?: string;
+  /**
+   * Visible text label for the toggle.
+   */
+  children?: React.ReactNode;
+  /**
+   * Visible text label for the toggle.
+   */
+  label?: React.ReactNode;
+};
+
+const ToggleLabel = ({ children, className }: ToggleLabelProps) => {
+  const componentClassName = clsx(styles['toggle__label'], className);
+  return <Switch.Label className={componentClassName}>{children}</Switch.Label>;
+};
+
+const ToggleButton = ({ className, checked, ...other }: ToggleButtonProps) => (
+  <Switch
+    checked={checked}
+    className={clsx(
+      styles['toggle__button'],
+      checked
+        ? styles['toggle__button--checked']
+        : styles['toggle__button--unchecked'],
+      className,
+    )}
+    {...other}
+  >
+    {/* This span is the circular part of the toggle */}
+    <span className={styles['toggle__button-circle']} />
+  </Switch>
+);
+
+const ToggleWrapper = Switch.Group;
+
+/**
+ * BETA: This component is still a work in progress and is subject to change.
+ *
+ * `import {Toggle} from "@chanzuckerberg/eds";`
+ *
+ * Toggle wrapping the Headless UI Switch component https://headlessui.dev/react/switch, generally used as an input for controlling between two states.
+ *
+ *
+ * @example
+ * <Toggle checked={checked} label="Lorem Ipsum" onChange={onChange} />
+ *
+ * @example
+ * <Toggle.Wrapper as="Fragment" className={styles.customWrapperStyles}>
+ *   <Toggle.Label> Some label </Toggle.Label>
+ *   <Toggle.Button onChange={onChange} checked={checked} className={customCssModulesClassname} />
+ * </Toggle.Wrapper>
+ *
+ */
+export const Toggle = ({ label, ...other }: ToggleProps) => {
+  // TODO-JL: Consider if (children) return <Switch.Group>{children}</Switch.Group>
```

Comment:
Should this component be improved for dev experience where we check for presence of `children` and it automatically detects custom Toggle use? i.e.:
right now:
```tsx
<Toggle.Wrapper>
  <Toggle.Button>
  <Toggle.Label>
</Toggle.Wrapper>
```
but with detection of children we can do
```tsx
<Toggle.Wrapper>
  <Toggle.Button>
  <Toggle.Label>
</Toggle.Wrapper>
```
via something like
```tsx
if (children)
return <ToggleWrapper>{children}</Toggle>
```

but the logic there regarding types gets complicated, and using `<ToggleWrapper>` not that big of a lift for consumers anyways, and probably discourages customization.

Lmk any thoughts

Reply:
I'm fine with `<Toggle.Wrapper>` 🤷 I can see it possibly being confusing at first because I might expect `Toggle.Wrapper` to wrap `Toggle`, like `<Toggle.Wrapper><Toggle></Toggle></Toggle.Wrapper>`, but I don't think it's too bad.